plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java

```diff
@@ -169,7 +200,103 @@ static TrinoException handleException(ParquetDataSourceId dataSourceId, Exceptio
         return new TrinoException(HIVE_CURSOR_ERROR, format("Failed to read Parquet file: %s", dataSourceId), exception);
     }
 
-    private static Block getRowIndexColumn(long baseIndex, int size)
+    private static boolean isColumnAdaptationRequired(List<ColumnAdaptation> columnAdaptations)
+    {
+        boolean isColumnAdaptationRequired = false;
+        // If no synthetic columns are added and the source columns are in order, no adaptations are required
+        for (int columnChannel = 0; columnChannel < columnAdaptations.size(); columnChannel++) {
+            ColumnAdaptation column = columnAdaptations.get(columnChannel);
+            if (column instanceof SourceColumn) {
+                int delegateChannel = ((SourceColumn) column).getSourceChannel();
+                if (columnChannel != delegateChannel) {
+                    isColumnAdaptationRequired = true;
+                    break;
+                }
+            }
+            else {
+                isColumnAdaptationRequired = true;
+                break;
+            }
+        }
+        return isColumnAdaptationRequired;
+    }
+
+    private interface ColumnAdaptation
+    {
+        Block block(Page sourcePage, long startRowId);
+    }
+
+    private static class NullColumn
+            implements ColumnAdaptation
+    {
+        private final Block nullBlock;
+
+        private NullColumn(Type type)
+        {
+            this.nullBlock = type.createBlockBuilder(null, 1, 0)
+                    .appendNull()
+                    .build();
+        }
+
+        @Override
+        public Block block(Page sourcePage, long startRowId)
+        {
+            return RunLengthEncodedBlock.create(nullBlock, sourcePage.getPositionCount());
+        }
+    }
+
+    private static class SourceColumn
+            implements ColumnAdaptation
+    {
+        private final int sourceChannel;
+
+        private SourceColumn(int sourceChannel)
+        {
+            checkArgument(sourceChannel >= 0, "sourceChannel is negative");
+            this.sourceChannel = sourceChannel;
+        }
+
+        @Override
+        public Block block(Page sourcePage, long startRowId)
+        {
+            return sourcePage.getBlock(sourceChannel);
+        }
+
+        private int getSourceChannel()
```

Comment:
I'd make this public. The class itself is private so it won't be visible outside.
A simple getter being private may confuse readers